Sentinel 2pc controller bug fixes

mszulcz-mitre opened the pull request:
This PR addresses the bugs identified in Issue #140 and Issue #141.  The bug fixes and associated unit tests are the same as those described in the issues.

mszulcz-mitre committed:
Fix bug that could cause non-existent element access

Previously, sentinel_2pc::controller would incorrectly calculate the
upper bound of the sentinel ID distribution if the number of sentinel
clients was 0.  The incorrect calculation was due to casting a negative
number to an unsigned int.  As a result of this error, the sentinel IDs
taken from the distribution could exceed the range of the possible
sentinel IDs, which could lead to a non-existent element access when a
bad ID was used as an index to a vector.

Signed-off-by: Michael L. Szulczewski <mszulczewski@mitre.org>

mszulcz-mitre committed:
Check precondition when initializing sentinel 2PC controller

sentinel_2pc::controller allows the sentinel ID to exceed the number of
sentinels, which causes a non-existent element access (see GitHub
Issue #140).  This commit checks for that condition and throws an error if
it's true.  It also adds a unit test to confirm that the code behaves
correctly after the fix.

Signed-off-by: Michael L. Szulczewski <mszulczewski@mitre.org>

metalicjames commented on src/uhs/twophase/sentinel_2pc/controller.cpp:
```diff
@@ -25,6 +25,20 @@ namespace cbdc::sentinel_2pc {
+            return false;
+        }
+
+        if(m_opts.m_attestation_threshold
```
I think this check more appropriately belongs in the configuration file parsing code where there are other sanity checks like this.

mszulcz-mitre replied:
Good call.  In the latest commit, I put it with the other sanity checks on the `options` struct in the function `cbdc::config::check_options`.

HalosGhost replied:
For my own clarity, which check got moved?

HalosGhost replied:
Pardon the noise. The way github is displaying this is quite strange.

mszulcz-mitre committed:
Check precondition in the options struct

Check that the number of required attestations doesn't exceed the number
of sentinels that can provide them.  Also add a unit test to confirm
that the check works.

Signed-off-by: Michael L. Szulczewski <mszulczewski@mitre.org>

HalosGhost commented on src/util/common/config.cpp:
```diff
@@ -682,6 +682,11 @@ namespace cbdc::config {
                 return "Two-phase mode requires at least one configured "
                        "sentinel";
             }
+            if(opts.m_sentinel_endpoints.size()
```
It looks like this commit added this check but didn't remove it from `::init()`; is it valuable having this check twice?

mszulcz-mitre replied:
Nope--it's probably not valuable.  I just forgot about it.  I just refactored the commits to remove it and to hopefully improve the commit messages.

HalosGhost commented:
@mszulcz-mitre this has a very minimal merge conflict with `trunk` (just the addition of some tests). When you have a moment and can rebase it, I'll give it one final review and we'll get this merged as well!

HalosGhost commented:
@metalicjames when you have a chance to review this, I'd love a second set of eyes.

metalicjames commented on src/uhs/twophase/sentinel_2pc/controller.cpp:
```diff
@@ -25,6 +25,12 @@ namespace cbdc::sentinel_2pc {
                                                    .size())]) {}
 
     auto controller::init() -> bool {
+        if(m_sentinel_id > (m_opts.m_sentinel_endpoints.size() - 1)) {
```
Does this check still work if `m_sentinel_endpoints.size()` is zero? Does `m_opts.m_sentinel_endpoints.size() - 1` result in underflow, bypassing the check?

metalicjames reviewed (commented):
I think we should add a unit test for the potential underflow case, then I think it's good to merge.

mszulcz-mitre committed:
Check that sentinel endpoints are defined

Previously, the init() method of the 2PC sentinel controller would
exhibit a non-existent element access if no sentinel endpoints were
defined.  With this commit, the code now checks that at least one
endpoint is defined and logs an error if it's not.

Signed-off-by: Michael L. Szulczewski <mszulczewski@mitre.org>

mszulcz-mitre commented:
I modified the code to avoid the possibility of underflow.  For clarity, I added another check to confirm that at least one sentinel endpoint is defined and I added a related unit test too.

mszulcz-mitre commented:
@HalosGhost It looks like James approved the changes I made in response to his review.  Would you like me to do anything more for this pull request?

HalosGhost commented:
Sorry for the delay; I've finally had a chance to walk through the code locally and test it. Tested ACK. All these changes eliminate several failure points and do so minimally. Looks good to me.
